### PR TITLE
Add support for modular build structure.

### PR DESCRIPTION
**Pull request by grafikrobot**
This is part of the effort to make the Boost libraries "modular" for build and consumption. See https://lists.boost.org/Archives/boost/2024/01/255704.php and https://github.com/grafikrobot/boost-b2-modular/blob/b2-modular/README.adoc for more information.

This PR depends on the following other PRs being merged to both develop and master branches of the respective repos:

- https://github.com/boostorg/boost/pull/854

This PR will be changed to ready for review, i.e. not draft, when the above are merged. Do not merge this one until that time.

### CODE REVIEW

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

File: test/build/Jamfile.v2 [hunk 1]
```diff
@@ -51,13 +70,18 @@ project wave/test
         cxx11_hdr_mutex
         cxx11_hdr_regex
       ]
+      <include>$(SAMPLES_DIR)
```
Comment (jefftrull) [inline test/build/Jamfile.v2:73]: I'm surprised to see that the tests might depend on the samples (and that this was detected in the process of this work). Can you recall why this is required?

Comment (jefftrull) [reply test/build/Jamfile.v2:73]: I see from later changes that it's `lexertl` :)

**Comment by jefftrull** [comment]
This is great work and I appreciate the effort you've put into making this possible. Before I merge, though, I wanted to ask about the current CI failures - will they clear up soon? Some sort of gcc version issue with a Docker container?

**Comment by jefftrull** [comment]
AFAICT the libc version issues in CI are  unrelated and affecting other Boost libs. I'm going to track them down first so I have a green build to start from.

**Pull request merged**